test/borda_clj/trace.clj

```diff
@@ -0,0 +1,196 @@
+                                [1 (s/gen ::values)]]))]
+            [1 (s/gen ::trace-input-entry)]]))))))))
+
+(s/def ::run-trace-args
+  (s/spec (s/cat :global-dimensions ::dimensions :max-buffer-size pos-int? :trace ::input-trace)
+          :gen gen-run-trace-args))
+
+(defn expect-some-output
+  "If we send any measurements, we expect some output"
+  [x]
+  (if (some #(= (first %) :measurement) (-> x :args :trace))
+    (not-empty (-> x :ret))
+    true))
+
+(defn map-in? [m1 m2]
+  (= m1 (select-keys m2 (keys m1))))
+
+(defn globals-are-in-all-sends [{:keys [ret] {:keys [global-dimensions]} :args}]
+  (every? (fn [{{measurements ::send} :send}]
+             (if (some? measurements)
+               (every?
+                (fn [k]
+                  (map-in? global-dimensions k))
+                (keys measurements))
+               true))
+          (seq ret)))
+
+(defn total-reported [global-dimensions trace]
+  (->> trace
+       (filter #(= (first %) :measurement))
+       (map second)
+       (reduce (fn [m [d v]]
+                 (update m (merge global-dimensions d) (partial merge-with +) v))
+               {})))
+
+(defn total-sent [global-dimensions result]
+  (->> result
+       (filter #(= (first %) :send))
+       (map second)
+       (map ::send)
+       (reduce (partial merge-with (partial merge-with +)))
+       (#(dissoc % (merge global-dimensions sut/submit-key)))))
+
+(defn sent<=reported
+  "The total quantity of successfully sent data in each dimension is less than
+  or equal than the total quantity we reported."
+  [{:keys [ret] {:keys [global-dimensions trace]} :args}]
+  (let [reported (total-reported global-dimensions trace)
+        sent (total-sent global-dimensions ret)]
+    (every? (fn [k]
+              (let [reported-vals (get reported k)
+                    sent-vals (get sent k)]
+                ;; XXX: this also passes if we change <= to =, which I suspect
+                ;; reflects a problem with our test data generator
+                (every? #(<= (% sent-vals) (% reported-vals))
+                        (keys sent-vals))))
+            (keys sent))))
+
+(defn delays-dont-affect-result
+  "Unfortunately this is not true. The output stream is not a deterministic
+  function of the input stream. In particular, ticks can see reports that
+  happened after them. This doesn't make our logic incorrect, only less
+  straightforward to test."
+  [x]
+  (= (->> x :ret (s/unform ::output-trace))
+     (run-trace (->> x :args :global-dimensions (s/unform ::dimensions))
+                (-> x :args :max-buffer-size)
+                (->> x :args :trace (s/unform ::input-trace) (remove #{:delay})))))
+
+(s/fdef
+ run-trace
+ :args ::run-trace-args
+ :ret ::output-trace
+ :fn (s/and expect-some-output
+            globals-are-in-all-sends
+            sent<=reported))
+
+
+(comment
+
+  (stest/abbrev-result (first (stest/check `run-trace {:clojure.spec.test.check/opts {:num-tests 1000 :max-size 5}})))
```

Comment:
This is what you'd call to exercise the auto-generated tests for the `run-trace` function, and therefore on much of the reporting logic.